src/ws/Socket.ts

```diff
@@ -461,16 +462,17 @@ export class Socket extends EventEmitter {
 
         // If we already authed, it means we're reconnecting and should
         // establish authentication again.
-        if (this._authpacket) {
+        if (this._authpacket || this._optOutEventsArgs.length) {
             // tslint:disable-next-line no-floating-promises
-            this.call(authMethod, this._authpacket, { force: true })
+            const promise = this._optOutEventsArgs.length
+                ? this.call('optOutEvents', this._optOutEventsArgs, { force: true }).catch(() =>
+                      this.emit('error', new UnknownCodeError()),
+                  )
+                : Promise.resolve();
+            promise
+                .then(() => this.emit('optOutResult'))
```

Comment:
this logic should be encapsulated in the `optOutEvents` method. It is beyond the responsibilities of `unspool`.
A much cleaner way would be to attach this logic to the `WelcomeEvent`

Reply:
This is similar to `auth` method and hence needs to be spooled and unspooled on reconnection.